Upgrade to Postgres 16

msepga opened the pull request:
TODO:

- [x] Fix remaining failing deparse tests
- [x] Fix unoptimized debug builds
- [x] Fix leaks detected by valgrind

msepga committed:
Parse auto-generated `nodetags.h` to get node types

msepga committed:
Include auto-generated NodeTag enum values

msepga committed:
Bump `PG_VERSION`

msepga committed:
Disable ICU during Postgres configuration

msepga committed:
Drop trailing junk removal patch

msepga committed:
Update patch offsets

msepga committed:
Exclude `*.switch.c` and `*.funcs.c` files

lfittl commented:
@msepga FWIW, I think the following log output when building may be worth investigating:

(1)
```
 In file included from src/pg_query_outfuncs_json.c:272:
src/pg_query_outfuncs_defs.c:1243:3: warning: format specifies type 'unsigned int' but the argument has type 'AclMode' (aka 'unsigned long') [-Wformat]
  WRITE_UINT_FIELD(required_perms, requiredPerms, requiredPerms);
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/pg_query_outfuncs_json.c:36:66: note: expanded from macro 'WRITE_UINT_FIELD'
                appendStringInfo(out, "\"" CppAsString(outname_json) "\":%u,", node->fldname); \
                                                                         ~~    ^~~~~~~~~~~~~
1 warning generated.
compiling src/pg_query_fingerprint.c
In file included from src/pg_query_fingerprint.c:263:
src/pg_query_fingerprint_defs.c:5897:27: warning: format specifies type 'int' but the argument has type 'AclMode' (aka 'unsigned long') [-Wformat]
    sprintf(buffer, "%d", node->requiredPerms);
                     ~~   ^~~~~~~~~~~~~~~~~~~
                     %lu
1 warning generated.
```

That looks like we maybe got the wrong type somewhere for AclMode - possibly that needs to be added somewhere in the extraction logic that is used as a source for generating those files.

(2)

```
compiling src/postgres_deparse.c
src/postgres_deparse.c:3286:10: warning: enumeration value 'JOIN_RIGHT_ANTI' not handled in switch [-Wswitch]
        switch (join_expr->jointype)
                ^
1 warning generated.
```

It appears JOIN_RIGHT_ANTI is only used by the planner, not during raw parsing, at least if I read https://git.postgresql.org/gitweb/?p=postgresql.git;a=commitdiff;h=16dc2703c5413534d4989e08253e8f4fcb0e2aab correctly. So I think that just needs to be added next to JOIN_ANTI in that case statement, to avoid the warning: https://github.com/pganalyze/libpg_query/blob/bd15fd5614d59ea92d9d6a53821f23d652bc940b/src/postgres_deparse.c#L3302

lfittl commented on Makefile:
```diff
@@ -6,7 +6,7 @@ ARLIB = lib$(TARGET).a
 PGDIRBZ2 = $(root_dir)/tmp/postgres.tar.bz2
 
-PG_VERSION = 15.1
+PG_VERSION = 16.0
```
Might be good to use `16.1` here, since that's out now.

msepga committed:
Fix deparsing of partition strategy

msepga committed:
Update `GRANT` option deparsing

msepga committed:
Ignore include-only C files in `OBJ_LIST`

msepga committed:
Use `bms_next_member` in `pg_query_outfuncs_protobuf.c`

msepga committed:
Add `RelFileNumber` as `INT_TYPE`

This is a typedef to `Oid`.

msepga committed:
Use `Assert` over deprecated `AssertArg`

Removed in upstream Postgres as part of b1099eca8f38ff5cfaf0901bb91cb6a22f909bc6

msepga committed:
Resolve new symbol `raw_expression_tree_walker_impl`

msepga committed:
Fix paths for gramparse.h

msepga committed:
Add gramparse as header extraction group

msepga committed:
Replace deprecated `bms_first_member` with `bms_next_member` in outfuncs

msepga committed:
Correctly deparse `REVOKE ... CASCADE`

msepga committed:
Remove old internal node cases when deparsing `ALTER TABLE`

msepga committed:
Update `REVOKE` option deparsing

msepga committed:
Parse `pg_node_attr` on struct fields

msepga committed:
Update parse test version numbers for Postgres 16.1

msepga committed:
Fix deparsing for `COPY ... FORMAT TEXT ...`

msepga committed:
Deparse `STORAGE` mode

msepga committed:
Bump Postgres version to 16.1

msepga committed:
Extract sources

msepga committed:
Fix `xml_serialize` deparsing

msepga committed:
Add parens around `A_Expr`s when used with `AT TIME ZONE`

msepga committed:
Generate fingerprint outfuncs

msepga committed:
Deparse `COERCE_SQL_SYNTAX` format `SYSTEM_USER` correctly

msepga committed:
Add case for `JOIN_RIGHT_ANTI`

msepga committed:
Update `AclMode` handling to be 64 bits wide

msepga committed:
Generate protobuf and funcs

msepga committed:
Extract headers

msepga committed:
Remove parse test for integer trailing junk

msepga committed:
Generate C protobuf bindings

msepga committed:
Update parse test version numbers

msepga committed:
Add patch for trailing junk on param references

msepga committed:
Fix debug builds

msepga commented:
The `AclMode` issue should be fixed in d2f9781d, and we now add a case for `JOIN_RIGHT_ANTI` in 6a47fa40.

I've updated the changes here to match Postgres 16.1 and resolved remaining test failures. It looks like there are still some valgrind issues to work out here before this is ready to merge.

msepga committed:
Replace use of mmap with plain read

We pass the file contents to the Postgres scanner, which assumes it is
given a string as input. As such, it uses `strlen`. `mmap` doesn't
provide us with a terminal null character, so we rely on empty pages to
provide this.

In the case of the `incremental_sort.sql` test, it is 12288 bytes, a
multiple of exactly 3 of the typical 4096 page size. As such, this test
ends up not getting *any* null terminators from `mmap`, so reading can
overrun the buffer.

Now, we just use a plain read and append `0` to the buffer to avoid the
problem.

msepga commented:
Looks like the valgrind issue was actually triggered by chance here; the new `incremental_sort.sql` test from upstream is exactly `12288` bytes, which is exactly 3 times the typical `4096` page size used with `mmap`. This means loading the test wasn't providing us with free null terminator bytes (typical when mmap zeros the rest of the page).

Patched in 397ccc9b.

lelit commented:
Hi @msepga, I'm starting to upgrade my pglast to this...

I see that the [extracted list of node tags](https://github.com/pganalyze/libpg_query/blob/16-latest-dev/srcdata/enum_defs.json#L2193) are missing the usual `T_` prefix: is that expected/on purpose?

msepga committed:
Fix `NodeTag` variant names

msepga commented:
> I see that the [extracted list of node tags](https://github.com/pganalyze/libpg_query/blob/16-latest-dev/srcdata/enum_defs.json#L2193) are missing the usual `T_` prefix: is that expected/on purpose?

Thanks for spotting that @lelit 🙂 Fixed in 3a64b99c & regenerated in f7160462.

seanlinsley commented:
I'm not sure if we have a process to keep the different branches in sync, but it looks like 16-latest is 3 commits behind 15-latest. This PR includes one of those commits https://github.com/pganalyze/libpg_query/compare/16-latest...15-latest

lelit commented:
I almost completed the upgrade of pglast (see its [v6 branch](https://github.com/lelit/pglast/tree/v6)) and all tests are green :partying_face: 

On the GH Actions, there's a [failure on the job that builds a 32bit wheel](https://github.com/lelit/pglast/actions/runs/7272884341/job/19815796355#step:5:1396): it's related to the fingerprinting functionality, does that ring any bell to you?

lfittl commented:
> I almost completed the upgrade of pglast (see its [v6 branch](https://github.com/lelit/pglast/tree/v6)) and all tests are green 🥳

Excellent, thanks for the testing!

FYI, I discussed the plan for releasing this PR with @msepga earlier today - the plan is to merge most of the open PRs into 15-latest first, then rebase this, and then make the actual 16-latest release at the end of the week.

Thus, I'd recommend not making an official release of the pglast v6 branch just yet, until we tagged a release here, just to avoid any unexpected changes when we do that rebase.

> On the GH Actions, there's a [failure on the job that builds a 32bit wheel](https://github.com/lelit/pglast/actions/runs/7272884341/job/19815796355#step:5:1396): it's related to the fingerprinting functionality, does that ring any bell to you?

Copying the error for easier context:

```
         In file included from ./src/postgres/include/access/htup_details.h:20,
                         from ./src/postgres/include/executor/tuptable.h:18,
                         from ./src/postgres/include/access/tupconvert.h:20,
                         from ./src/postgres/include/nodes/execnodes.h:32,
                         from ./src/postgres/include/catalog/index.h:18,
                         from src/postgres_deparse.c:2:
        ./src/postgres/include/access/tupmacs.h: In function ‘fetch_att’:
        ./src/postgres/include/access/tupmacs.h:65:4: error: duplicate case value
           65 |    case sizeof(Datum):
              |    ^~~~
        ./src/postgres/include/access/tupmacs.h:62:4: note: previously used here
           62 |    case sizeof(int32):
              |    ^~~~
```

This looks to me like that 32-bit build is using the wrong setting for SIZEOF_DATUM - we don't really support 32-bit builds, but if you wanted to make them work, you'd have to override the SIZEOF_VOID_P and other size-related flags defined in `pg_config.h` (these are usually set by autoconf / meson in regular Postgres, but in case of libpg_query we hardcode them).

I don't think that's a new problem on the 16 branch, though it could be that for some reason this worked by coincidence in earlier versions. For now I'd suggest we discuss 32-bit issues separately from this PR.

lelit commented:
> FYI, I discussed the plan for releasing this PR with @msepga earlier today - the plan is to merge most of the open PRs into 15-latest first, then rebase this, and then make the actual 16-latest release at the end of the week.
> 
> Thus, I'd recommend not making an official release of the pglast v6 branch just yet, until we tagged a release here, just to avoid any unexpected changes when we do that rebase.

Sure, there's no hurry.

> > On the GH Actions, there's a failure...
> 
> This looks to me like that 32-bit build is using the wrong setting for SIZEOF_DATUM - we don't really support 32-bit builds, but if you wanted to make them work, you'd have to override the SIZEOF_VOID_P and other size-related flags defined in `pg_config.h` (these are usually set by autoconf / meson in regular Postgres, but in case of libpg_query we hardcode them).

Thank you, I will try to understand what's going on.

> I don't think that's a new problem on the 16 branch, though it could be that for some reason this worked by coincidence in earlier versions. For now I'd suggest we discuss 32-bit issues separately from this PR.

Ok, will report back somewhere else if I find something interesting.

msepga committed:
Update version in parse opt tests

msepga committed:
Regenerate C protobuf definitions with `protoc 25.1`

msepga committed:
Bump `PROTOC_VERSION` to 25.1, use C++17 for C++ builds

msepga committed:
Regenerate C++ protobuf definitions with `protoc 25.1`

msepga committed:
Bump actions workflow to use protobuf 25.1

msepga committed:
Link to `-lm`

msepga committed:
Suppress more leaks related to unclean protobuf shutdown

lfittl reviewed (approved):
✅ I've tested this, reviewed the files to make sure we don't pull in unnecessary code, and confirmed that the protobuf compiler now (after the update to 25.1) produces the exact `protobuf/` folder contents for me locally.

Nicely done 🥳 

Also, FYI, I pushed up some more deparser fixes in https://github.com/pganalyze/libpg_query/pull/229 -- we could merge those in separately after this is merged, or you can pull this into this PR, as you prefer.

lfittl commented:
> > This looks to me like that 32-bit build is using the wrong setting for SIZEOF_DATUM - we don't really support 32-bit builds, but if you wanted to make them work, you'd have to override the SIZEOF_VOID_P and other size-related flags defined in `pg_config.h` (these are usually set by autoconf / meson in regular Postgres, but in case of libpg_query we hardcode them).
> 
> Thank you, I will try to understand what's going on.
> 
> > I don't think that's a new problem on the 16 branch, though it could be that for some reason this worked by coincidence in earlier versions. For now I'd suggest we discuss 32-bit issues separately from this PR.
> 
> Ok, will report back somewhere else if I find something interesting.

@lelit FWIW, it appears there was actually a new issue on 16, which is that we were missing some of the functions required when building with 32-bit.

I ended up putting a bit of effort into this, so that we can actually support 32-bit out of the box with some extra define/undefs + extracting a bit more code. See https://github.com/pganalyze/libpg_query/pull/232
